### DISCUSSION TITLE
langgraph checkpoint

### DISCUSSION THREAD

**Discussion by wenliangz** [JoshuaC215/agent-service-toolkit#29] [Q&A] [answered]
very nice work for langgraph agents! One thing I nocitced that the checkpoint doesn't seem to work as expected. We lose the context between multiple questions. Does somebody else have the same problem?

**Comment by JoshuaC215** [comment]
Hi @wenliangz thank you!

I just want to make sure you’re sending the same thread_id for subsequent requests and keeping the server / container alive?

I haven’t had trouble with this recently (you can see the deployed agent linked in the readme maintains context within the same chat, which uses the checkpoint mechanism)

Feel free to post more info about your setup or repro steps if useful and happy to see if I can help

**Reply by wenliangz** [reply]
Thank you for your quick response! I did notice that your deployed agent has memory working. I am in the process of trouble shooting now. I have some modifications of your code, but I do believe that I have the same thread_id, which is the st session_id. And I also have the checkpoint in place. I run the app in the dockers. Is there any other places that might go wrong? Thanks again!

**Reply by JoshuaC215** [reply]
Ok, interesting, I’m not totally sure. Things to try:
* use the SQLite client to inspect the checkpoint database contents and see if it’s what you expect
* Add logging to the service calls to ensure the expected values are being passed to LangGraph
* See if LangGraph Studio or LangSmith provide any better visibility into what’s happening

**Reply by wenliangz** [reply]
Thank you very much for your advice! In my case, we are using our own agent, which is quite complicated. However, we adapted your front-end. I am investigating right now why my agent has lost the memory even we have the checkpoints and the same thread_id. One thing I notice in your agent, your entry node, has the prompt function to include the chat history. We don't have this. I thought the history will be automatically added when we apply the checkpoint? I am not sure exactly how the checkpoint work at the backend. Any advice is appreciated!  
```
    def _compile_prompt(self, role: str, messages: List[AnyMessage]) -> str:
        role_mapping = {"ai": "Agent", "human": "User"}
        messages_str = [
            f"{role_mapping[m.type]}: {m.content}" for m in messages if m.type in ["ai", "human"]
        ]
        conversation_history = "\n\n".join(messages_str)
        return self.prompt.format(role=role, conversation_history=conversation_history)
```

**Reply by JoshuaC215** [reply, marked as answer]
Good question - the code you referenced (from [here](https://github.com/JoshuaC215/agent-service-toolkit/blob/0c5d00600666cfc1814eee075414b6737623a5e7/src/agent/llama_guard.py#L86-L92)) is used to send a conversation summary to LlamaGuard only for content moderation BEFORE the actual agent evaluation, so it shouldn't be impacting the result. Arguably, it's overkill and it would be OK to just send the last message here.

For the "main" agent evaluation, note that conversation history is stored in `state["messages"]` and when you use an existing thread_id, this value is pre-populated with the earlier messages. You are still responsible for taking those messages and sending them to the LLM - this is how LangGraph works, not specific to this repo. In my example agent, that part happens [here](https://github.com/JoshuaC215/agent-service-toolkit/blob/0c5d00600666cfc1814eee075414b6737623a5e7/src/agent/research_assistant.py#L55-L74).
